Exercise: Clean code & MPV pattern

herrerameri opened the pull request:
- Rename variables and methods
- Dividing the code into shorter and focused submethods 
- Handling error
- Create custom Toast for user messages
- MVP pattern 

![cleaning](https://cloud.githubusercontent.com/assets/22835487/19515668/add2c8d6-95f8-11e6-8a37-3b63832db90a.gif)

herrerameri committed:
Cleaning the code: variables rename, refactor, deleting comments lines, dividing the code in submethods, focused.

herrerameri committed:
Cleaning code: add new files

herrerameri committed:
Implementing MVP pattern.

Akelael commented on app/src/main/java/edu/akelael/comics/model/ComicListModel.java:
```diff
@@ -0,0 +1,48 @@
+public class ComicListModel implements IComicListModel {
+    Server server;
+    Retrofit retrofit;
+    HashMap<String, String> mMap;
```
I believe these three attributes doesn't need to be package private. Their visibility can be made more restrictive probably. Also, IMO their names doesn't tell what they are for. mMap is map, that's something that can be seen in the declaration, but what it really improves readability is to explain what is used for. For example, I would name it something like "apiRequestContent", as it includes the timestamp, api key and hash for making a request to the Marvel API.

On the other hand, the hungarian notation (that includes adding an m before the name of a field) was used when we didn't have syntax highlights in our IDEs and it helped the programmers to realize that was a field. Old projects can still use it as a part of their legacy (code), but now is completely useless, and it worsens legibility.

herrerameri replied:
Yes! I know about de hungarian notation, I will correct it! Thanks!

Akelael commented on app/src/main/java/edu/akelael/comics/model/ComicListModel.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public void getComics(final IComicListModelCallbacks listener){
+        server = retrofit.create(Server.class);
```
This call can be moved to the initializeApiConnection method, so it will be called only once. It makes sense IMO, as now a Server object is created each time you retrieve the comics list, and this movement can considered as a part of the api initialization process. Also, if what I propose is done, there's no need for the "retrofit" field to be a field anymore (it can be made a local variable).

herrerameri replied:
IMO here the method has its own purpose that is "get the comics". In fact, for me it's a candidate to be called from another piece of code, so is better to have it in there with its responsability. 
I think it's not clear to put this on the "initializeApiConnection()" cause also semantically it has other meaning.

Akelael commented on app/src/main/java/edu/akelael/comics/model/ComicListModel.java:
```diff
@@ -0,0 +1,48 @@
+        mMap = new HashMap<>();
+        mMap.put("ts", "ts");
+        mMap.put("apikey", "apikey");
+        mMap.put("hash", "hash");
```
I see 2 different things in this method (though both of them make sense to be here IMO). First, there is the creation of the "retrofit" object (or server object, if the suggestion I made before is applied). Then, there is the initialization of the hash map with the request content. What I believe it would improve readability in this case is to move these two things to two separate methods. That way, the initializeApiConnection could be read something like:

``` java
public void initializeApiConnection() {
    intializeServer(); // not the best name IMO
    initializeRequestContent();
}
```

herrerameri replied:
Yes, may be is cleaner... I will correct it!

Akelael commented on app/src/main/java/edu/akelael/comics/model/ComicListModel.java:
```diff
@@ -0,0 +1,48 @@
+        marvelCall.enqueue(new Callback<Marvel>() {
+            @Override
+            public void onResponse(Call<Marvel> call, Response<Marvel> response) {
+                if (response.code() == 200) {
```
200 is what we call a "magic number". Someone reading this code might invest a lot of time trying to figure what the author tried to do (even if the author is you 3 months ago). In this case 200 is the OK code for an HTTP response, so we could make it a constant and name it RESPONSE_OK, for example. For this kind of refactor in Android Studio, you can position the cursor in the variable, and use Cmd + Alt + C in a Mac  (or Ctrl + Alt + C in a Windows).

herrerameri replied:
Hahaha it's like an obvious number, not magic! 200 is the number all programmers wanna get from a response... we know it! Ok, I will correct it!

Akelael commented on app/src/main/java/edu/akelael/comics/model/IComicListModel.java:
```diff
@@ -0,0 +1,6 @@
+package edu.akelael.comics.model;
+
+public interface IComicListModel {
```
Using and I to start an interface name, or using Impl for a class that implements an interface, are usually smells. In MVP code examples, interfaces are usually provided for abstraction, but do they really are adding that value?

herrerameri replied:
I think yes, but in this exercise it's difficult to distinguish the correct use of the interfaces...

Akelael commented on app/src/main/java/edu/akelael/comics/model/Marvel.java:
```diff
@@ -1,9 +1,9 @@
 import java.util.List;
 
-class Marvel {
+public class Marvel {
```
IMO this Marvel name doesn't explain what this class really represent.

herrerameri replied:
Ohhh! Yes, I forgot to change it!! Sorry! I will correct it! Thanks!

Akelael commented on app/src/main/java/edu/akelael/comics/model/Server.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface Server {
+    String baseUrl = "http://gateway.marvel.com/v1/public/";
+    int ID_AMAZING_SPIDERMAN = 1010733;
```
A constant in Java is made by adding static and final modifiers.

herrerameri replied:
Yes... but for interfaces isn't it redundant? But clean... Mmm I don't know what is the best decision. I think that with the modifiers is cleaner.

Akelael commented on app/src/main/java/edu/akelael/comics/presenter/ComicListPresenter.java:
```diff
@@ -0,0 +1,35 @@
+import edu.akelael.comics.model.Marvel;
+import edu.akelael.comics.view.IComicListActivity;
+
+public class ComicListPresenter implements IComicListPresenter, IComicListModelCallbacks {
```
IMO this class has more responsibilities than it should have. If this is a presenter, then there's no need for it to act also as a callback. I suggest using an anonymous object for the callback instead of making the presenter implement the callback interface.

herrerameri replied:
I really like your idea...! 
Can you help me? Because I can't put the idea in code... 
This case is like this: VIEW needs COMICS, so PRESENTER tells MODEL "get me the comics". That is an async method.
When data is retrieved from server... MODEL has to say to PRESENTER "ok here is the data"? How without a callback? Sorry, I can't get it.... :(

herrerameri replied:
this was my idea... or intention...
![diagramamvp](https://cloud.githubusercontent.com/assets/22835487/20079324/71ec50ac-a544-11e6-8fbd-dc29f1750208.png)

Akelael commented on app/src/main/java/edu/akelael/comics/presenter/ComicListPresenter.java:
```diff
@@ -0,0 +1,35 @@
+
+public class ComicListPresenter implements IComicListPresenter, IComicListModelCallbacks {
+    IComicListActivity view;
+    IComicListModel model;
```
IMO the visibility of these fields could be more restrictive.

herrerameri replied:
It's corrected! Thanks!

herrerameri committed:
Corrections to PR#1:
- Change name of 'mMap' to 'apiRequestContent' in ComicListModel.java
- Change name of 'retrofit' to 'httpClient' in ComicListModel.java
- Create a constant for 200 HTTP code
- Divide method 'initializeApiConnection'
- Rename 'Marvel.java' to 'ComicData.java'
- Put 'final static' modifiers to 'ID_AMAZING_SPIDERMAN' and 'baseUrl'
- Put 'private' modifier to 'view' and 'model' in ComicListPresenter.java
